### ISSUE TITLE
How can I export textured mesh?

### ISSUE THREAD

**Issue by wukailu** [bennyguo/instant-nsr-pl#34]
The exported `.obj` file contains only the shape, not the corresponding material and color.

**Comment by bennyguo** [comment]
Hi, I haven't write code about this. But you could achieve this by (1) query the color (at a fixed viewing direction) for each mesh vertex; (2) rewrite the obj saving code to support color attributes. I'll add this feature when i have time.

**Comment by adithya1-7** [comment]
@bennyguo I was not able to figure [this](https://github.com/bennyguo/instant-nsr-pl/issues/28#issuecomment-1342306505) part out. Can you tell us which python files have to be modified to achieve the required output? It would be ideal if you provide the code modification snippets with line number. Thanks in advance

**Comment by iraj465** [comment]
@bennyguo getting textured mesh would be great! Can you please help on this?

**Comment by adithya1-7** [comment]
@bennyguo Found this snippet for colour extraction, but don't know how it applies to current implementation
`ray_proj = torch.from_numpy(vertices).cuda().float().split(8192 * 32, 0)`
`vertex_col = [model.mesh_color_forward(ray) for ray in ray_proj]`
`vertex_col = (torch.concat(vertex_col).cpu().detach().numpy() * 255.).astype( np.uint8)`
`mesh = trimesh.Trimesh(vertices, triangles, vertex_colors=vertex_col)`

**Comment by iraj465** [comment]
I was getting some idea from this:
https://github.com/xrr-233/Textured-NeuS

But not being able to implement those functions in this repo.

**Comment by bennyguo** [comment]
Hi guys, I just implemented the textured mesh export, please have a try! If you are resuming from old checkpoints, remember to add `export.chunk_size=2097152 export.export_vertex_color=true` after your command. For NeRF, you may need a higher marching cubes resolution (~512) to get better vertex color predictions.

**Comment by iraj465** [comment]
Getting this error. I turned on the flag for export mesh colour. This was working on the last commit. (just the mesh)
<img width="530" alt="Screenshot 2023-04-05 at 10 23 22 AM" src="https://user-images.githubusercontent.com/32464366/229984414-c8841b79-47c6-4a7d-a29c-5440952e86b1.png">

**Comment by bennyguo** [comment]
@iraj465 I also encountered this problem when viewing the mesh using MeshLab, but it seems just a warning. It may be related to using Trimesh to save the mesh. I'll look into it later today.

**Comment by iraj465** [comment]
Also, no mesh is coming in obj. Only a bunch of points, whereas it is being trained (in iters images)

Mesh obj:
<img width="1663" alt="Screenshot 2023-04-05 at 12 29 39 PM" src="https://user-images.githubusercontent.com/32464366/230005085-3f4d83c6-9580-48f6-bfaf-c89508fb703d.png">


Iters sample image:


![it5000-0](https://user-images.githubusercontent.com/32464366/230005182-3a83c433-2afc-4f99-8a00-6997e02e457c.png)

**Comment by bennyguo** [comment]
Are you training NeRF model? You need to tune `model.geometry.isosurface.threshold` to get reasonable meshes (in your case set to a smaller value).

**Comment by iraj465** [comment]
This is what i'm getting for the above shoe now.
<img width="916" alt="Screenshot 2023-04-05 at 2 37 11 PM" src="https://user-images.githubusercontent.com/32464366/230035729-1bbefa12-bcb8-4901-a2c9-7c172acf6cee.png">

Earlier i was getting this for the same threshold:?
<img width="695" alt="Screenshot 2023-04-05 at 2 40 25 PM" src="https://user-images.githubusercontent.com/32464366/230036225-b837a915-61a1-4a12-90a8-5812f0710ca8.png">

**Comment by bennyguo** [comment]
Any luck trying a lower threshold?

**Comment by bennyguo** [comment]
@iraj465 I think this is related to the new scene center estimation method mentioned in https://github.com/bennyguo/instant-nsr-pl/issues/55#issuecomment-1499944616. In your case it seems the scene center is biased, and your object is not fully wrapped by the bounding box. I'll push a fix later today.

**Comment by iraj465** [comment]
I see, the poses are not transformed relatively so the whole scene is shifted. Yeah @bennyguo the fix push would be great. I have also outlined this problem in this issue if you need further details or some testing to do. Issue #60

**Comment by bennyguo** [comment]
Camera normalization fixed in the latest commit.
